_source/_posts/2022-08-09-spring-boot-saml.adoc

```diff
@@ -0,0 +1,574 @@
+            singlelogout:
+              binding: POST
+              response-url: "{baseUrl}/logout/saml2/slo"
+----
+
+Restart your app, and you should be able to log in with Auth0.
+
+image::{% asset_path 'blog/spring-boot-saml2/auth0-login.png' %}[alt=Auth0 log in success,width=800,align=center]
+
+You might notice that the email and authorities are not calculated correctly. This is because the claim names have changed with Auth0. Update `SecurityConfiguration#groupsConverter()` to allow both Okta and Auth0 names for groups.
+
+[source,java]
+----
+private Converter<OpenSaml4AuthenticationProvider.ResponseToken, Saml2Authentication> groupsConverter() {
+
+    ...
+
+    return (responseToken) -> {
+        ...
+        List<String> groups = principal.getAttribute("groups");
+        // if groups is not preset, try Auth0 attribute name
+        if (groups == null) {
+            groups = principal.getAttribute("http://schemas.auth0.com/roles");
+        }
+        ...
+    };
+}
+----
+
+To make Auth0 populate a user's groups, navigate to *Auth Pipeline* > *Rules* and create a new Rule. Choose the Empty rule template. Provide a meaningful name like `Groups claim`, replace the `Script` content with the following, and *Save*.
+
+[source,js]
+----
+function(user, context, callback) {
+  user.preferred_username = user.email;
+  const roles = (context.authorization || {}).roles;
+
+  function prepareCustomClaimKey(claim) {
+    return `${claim}`;
+  }
+
+  const rolesClaim = prepareCustomClaimKey('roles');
+
+  if (context.idToken) {
+    context.idToken[rolesClaim] = roles;
+  }
+
+  if (context.accessToken) {
+    context.accessToken[rolesClaim] = roles;
+  }
+
+  callback(null, user, context);
+}
+----
+
+Next, modify `HomeController` to allow Auth0's email attribute name.
+
+[source,java]
+----
+public class HomeController {
+
+    @RequestMapping("/")
+    public String home(@AuthenticationPrincipal Saml2AuthenticatedPrincipal principal, Model model) {
+        model.addAttribute("name", principal.getName());
+        String email = principal.getFirstAttribute("email");
+        // if email is not preset, try Auth0 attribute name
+        if (email == null) {
+            email = principal.getFirstAttribute("http://schemas.xmlsoap.org/ws/2005/05/identity/claims/emailaddress");
+        }
+        model.addAttribute("emailAddress", email);
+        model.addAttribute("userAttributes", principal.getAttributes());
+        return "home";
+    }
+
+}
+----
+
+Restart your app, log in, and everything should work as expected.
+
+image::{% asset_path 'blog/spring-boot-saml2/auth0-groups.png' %}[alt=Auth0 groups,width=800,align=center]
+
+== Support Okta and Auth0
+
+You can also support _both_ Okta and Auth0! Modify your `application.yml` to be as follows, and you'll be prompted by Spring Security for which one to log in with. The `&name` and `*name` values are used to set and retrieve blocks of YAML to avoid repetition.
```

Comment:
```suggestion
You can also support _both_ Okta and Auth0! Modify your `application.yml` to be as follows, and Spring Security will prompt you for which one to log in with. The `&name` and `*name` values are used to set and retrieve blocks of YAML to avoid repetition.
```
changes passive voice to active